certora/specs/GuardianStorage.spec

```diff
@@ -0,0 +1,372 @@
+/*
+ * Spec for linked list reachability for Guardians.
+ * This file is derived from OwnerReach.spec for a Safe account.
+ * https://github.com/safe-global/safe-smart-account/blob/main/certora/specs/OwnerReach.spec
+ *
+ * This file uses a reach predicate:
+ *    ghost reach(address, address, address) returns bool
+ * to represent the transitive of the next
+ * relation given by the "owners" field relation for a specific wallet.
+ *
+ * The idea comes from the paper
+ *
+ * [1] Itzhaky, S., Banerjee, A., Immerman, N., Nanevski, A., Sagiv, M. (2013).
+ *     Effectively-Propositional Reasoning about Reachability in Linked Data Structures.
+ *     In: CAV 2013. Springer, https://doi.org/10.1007/978-3-642-39799-8_53
+ */
+
+using SafeHarness as safeContract;
+using SocialRecoveryModule as socialRecoveryModule;
+
+methods {
+    // Safe Functions
+    function safeContract.isModuleEnabled(address) external returns (bool) envfree;
+
+    // GuardianManager Functions
+    function threshold(address) external returns (uint256) envfree;
+    function isGuardian(address, address) external returns (bool) envfree;
+
+}
+
+persistent ghost reach(address, address, address) returns bool {
+    init_state axiom forall address X. forall address Y. forall address wallet. reach(wallet, X, Y) == (X == Y || to_mathint(Y) == 0);
+}
+
+persistent ghost mapping(address => mapping(address => address)) ghostGuardians {
+    init_state axiom forall address wallet. forall address X. to_mathint(ghostGuardians[wallet][X]) == 0;
+}
+
+persistent ghost ghostSuccCount(address, address) returns mathint {
+    init_state axiom forall address wallet. forall address X. ghostSuccCount(wallet, X) == 0;
+}
+
+persistent ghost mapping(address => uint256) ghostGuardianCount {
+    init_state axiom forall address X. to_mathint(ghostGuardianCount[X]) == 0;
+}
+
+persistent ghost address SENTINEL {
+    axiom to_mathint(SENTINEL) == 1;
+}
+
+persistent ghost address NULL {
+    axiom to_mathint(NULL) == 0;
+}
+
+// threshold can be 0 in case no Guardian is set for a wallet.
```

Comment:
Like we do in the other spec, can we have a description of each invariant and what it does? This comment and `thresholdSet` does not give much context.

What I understand from the code is that this is an "invariant that proves that the threshold is always lower than the guardian count".

Also, I think it is important to check that `threshold(wallet) == 0 => ghostGuardianCount[wallet] == 0`.

Reply:
added an invariant: https://github.com/5afe/CandideWalletContracts/pull/20/commits/a958a95dbc105937c1d2e5dfc7efd5d18c320db9